app/daemon_portforward.go

```diff
@@ -0,0 +1,227 @@
+
+type MenderPortForwarder struct {
+	ConnectionID string
+	SessionID    string
+	conn         net.Conn
+	ctx          context.Context
+	ctxCancel    context.CancelFunc
+}
+
+func (f *MenderPortForwarder) Connect(protocol string, host string, portNumber int) error {
+	log.Debugf("[sessionID:%s connectionID:%s] port-forwarding %s/%d", f.SessionID, f.ConnectionID, protocol, portNumber)
+
+	if protocol == protocolTCP || protocol == protocolUDP {
+		conn, err := net.Dial(protocol, host+":"+strconv.Itoa(portNumber))
+		if err != nil {
+			return err
+		}
+		f.conn = conn
+	} else {
+		return errors.New("unknown protocol: " + protocol)
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	f.ctx = ctx
+	f.ctxCancel = cancelFunc
+
+	go f.Read()
+
+	return nil
+}
+
+func (f *MenderPortForwarder) Close() error {
+	log.Debugf("[sessionID:%s connectionID:%s] close", f.SessionID, f.ConnectionID)
+	f.ctxCancel()
+	return f.conn.Close()
+}
+
+func (f *MenderPortForwarder) Read() {
+	errChan := make(chan error)
+	dataChan := make(chan []byte)
+
+	go func() {
+		data := make([]byte, readBuffLength)
+
+		for {
+			n, err := f.conn.Read(data)
+			if err != nil {
+				errChan <- err
+				break
+			}
+			if n > 0 {
+				dataChan <- data[:n]
+			}
+		}
+	}()
+
+	for {
+		select {
+		case err := <-errChan:
+			if err != io.EOF {
+				log.Errorf("[sessionID:%s connectionID:%s] error: %v\n", f.SessionID, f.ConnectionID, err.Error())
+			}
+			// send the stop message
+			m := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:     ProtoTypePortForward,
+					MsgType:   MessageTypePortForwardStop,
+					SessionID: f.SessionID,
+					Properties: map[string]interface{}{
+						propertyConnectionID: f.ConnectionID,
+					},
+				},
+				Body: []byte{},
+			}
+			_ = connectionmanager.Write(ws.ProtoTypeShell, m)
+			// close the forwarder
+			f.Close()
+		case data := <-dataChan:
+			m := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:     ProtoTypePortForward,
+					MsgType:   MessageTypePortForward,
+					SessionID: f.SessionID,
+					Properties: map[string]interface{}{
+						propertyConnectionID: f.ConnectionID,
+					},
+				},
+				Body: data,
+			}
+			_ = connectionmanager.Write(ws.ProtoTypeShell, m)
+		case <-f.ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *MenderPortForwarder) Write(body []byte) error {
+	_, err := f.conn.Write(body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getConnectionIDFromMessage(message *ws.ProtoMsg) string {
+	connectionID, _ := message.Header.Properties[propertyConnectionID].(string)
+	return connectionID
+}
+
+func (d *MenderShellDaemon) routeMessagePortForwardNew(message *ws.ProtoMsg) error {
+	connectionID := getConnectionIDFromMessage(message)
+	protocol := message.Header.Properties[propertyProtocol].(string)
+	host := message.Header.Properties[propertyRemoteHost].(string)
+	portNumber := message.Header.Properties[propertyRemotePort].(uint64)
```

Comment:
All these three rely on external network input and can crash the daemon if the type doesn't match.

Reply:
You are right, fixed!